modules/storage/storage.go

```diff
@@ -0,0 +1,176 @@
+package storage
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+
+	"gocloud.dev/blob"
+
+	// Google, Azure and S3 packages for bucket storage
+	_ "gocloud.dev/blob/azureblob"
+	_ "gocloud.dev/blob/fileblob"
+	_ "gocloud.dev/blob/gcsblob"
+	_ "gocloud.dev/blob/s3blob"
+)
+
+// FileStorage contains necessary info for reading or writing file to bucket
+type FileStorage struct {
+	Ctx      context.Context
+	Path     string
+	FileName string
+}
+
+// NewReader provides file reader from bucket and error if occurs
+func (fs *FileStorage) NewReader() (*blob.Reader, error) {
```

Comment:
If you pass an enum here, it's easy to decide between different file storages.

Reply:
If we want to separate storage system for avatars, lfs, attachments ( `[storage.avatars]`, `[storage.attachments]`, `[storage.lfs]` ), we  can use an extra field `BucketURL` here which will be useful to decide between file storages. 
https://github.com/go-gitea/gitea/blob/98c0c8ea2e0d4896cc3cc00423f0d2952fcbad21/modules/storage/storage.go#L22-L26